library/library.ml

```diff
@@ -278,48 +278,45 @@ let warn_several_object_files =
                             strbrk " instead of " ++ str vo ++
                             strbrk " because it is more recent")
 
-let locate_absolute_library dir =
-  (* Search in loadpath *)
-  let pref, base = split_dirpath dir in
-  let loadpath = Loadpath.filter_path (fun dir -> DirPath.equal dir pref) in
-  let () = match loadpath with [] -> raise LibUnmappedDir | _ -> () in
-  let loadpath = List.map fst loadpath in
-  let find ext =
-    try
-      let name = Id.to_string base ^ ext in
-      let _, file = System.where_in_path ~warn:false loadpath name in
-      Some file
-    with Not_found -> None in
-  match find ".vo", find ".vio" with
-  | None, None -> raise LibNotFound
-  | Some file, None | None, Some file -> file
-  | Some vo, Some vi when Unix.((stat vo).st_mtime < (stat vi).st_mtime) ->
-    warn_several_object_files (vi, vo);
-    vi
-  | Some vo, Some _ -> vo
-
-let locate_qualified_library ?root ?(warn = true) qid =
-  (* Search library in loadpath *)
-  let dir, base = repr_qualid qid in
-  let loadpath = Loadpath.expand_path ?root dir in
-  let () = match loadpath with [] -> raise LibUnmappedDir | _ -> () in
+type require_mode = RequireOnlyVo | RequireMostRecent
+
+let locate_library ~warn ~mode paths base =
+  let () = match paths with [] -> raise LibUnmappedDir | _ -> () in
   let find ext =
     try
       let name = Id.to_string base ^ ext in
       let lpath, file =
-        System.where_in_path ~warn (List.map fst loadpath) name in
+        System.where_in_path ~warn (List.map fst paths) name in
       Some (lpath, file)
     with Not_found -> None in
-  let lpath, file =
-    match find ".vo", find ".vio" with
+  match mode with
+  | RequireOnlyVo ->
+    begin match find ".vo" with
+      | None -> raise LibNotFound
+      | Some vo -> vo
+    end
+  | RequireMostRecent ->
+    begin match find ".vo", find ".vio" with
     | None, None -> raise LibNotFound
     | Some res, None | None, Some res -> res
     | Some (_, vo), Some (_, vi as resvi)
       when Unix.((stat vo).st_mtime < (stat vi).st_mtime) ->
       warn_several_object_files (vi, vo);
-       resvi
+      resvi
     | Some resvo, Some _ -> resvo
-  in
+    end
+
+let locate_absolute_library ~mode dir =
+  (* Search in loadpath *)
```

Comment:
Is the comment true? The name of the function seems to suggest that there is no search, while the code does a limited one.

Reply:
> The name of the function seems to suggest that there is no search

Why? Can you elaborate? "locate" seems like the description of a limited search to me, no?

Reply:
Below you have another `locate_` function with a very similar comment.
The two comments should tell the difference.

Reply:
Well, as their name indicates, one takes an absolute name, and the other one a qualid. Isn't it clear from both the types and the names?

Note that this PR doesn't introduce any of this.

Reply:
Oops, then OK. leave it like that.

What puzzles me as a reader is that if a "path" is absolute, then I don't get why some "location" is needed.
If any, this "location" is a different from the "location" that resolve a, say, relative path to an absolute one. It is more like the location that resolve a file name to sectors and cylinders. You are not the author of the mess (mixing the two concepts).
If it is clear to you what lines 312-313 actually do, then please comment it (and prune the two silly comments).